Autofill save

LongCatIsLooong opened the pull request:
## Description

Autofill save is the most common way to enable autofill for input fields of a specific "input content type", as it teaches the autofill service how to fill an input field of that content type in the future. 

### How it works on native platforms (Oversimplified)

**iOS** only saves usernames and passwords AFAIK. From what I'm seeing, every time an input field is removed from the view hierarchy, iOS keeps it in memory if it's a username or password field. If the field's "partner in crime" input field has already been removed (i.e. both username and password are now removed), iOS releases these fields from the memory after saving their text values.

On **Android**, one `AutofillManager` is created for each `Activity`, and each `AutofillManager` manages one `Session` at a time (thus each `Activity` can have only one `Session` at a time). The `Session` keeps track of the values of the fields that the autofill service expressed interested in saving during autofill. When an `Activity` is finished (implicitly destroys the `Session`) or a `Session` is manually destroyed, the values of these tracked fields may get sent to the autofill service for saving. 

On **web**, the input fields in a `form` may get saved when it's submitted.

### Problems this PR is trying to solve

On Android, since a flutter app only has one `Activity`, new autofillable input fields will be added to the same `Session` until the `Session` is destroyed. The system autofill manager may consume much more memory than needed in case the developer forgets to destroy the context when it's no longer needed. The same problem exists on web and iOS if we were to introduce autofill saving to these platforms. Native platforms more or less manage the lifespan of autofill contexts automatically.

### How it works

The platform text input plugin (in the case of Android, the system autofill manager service) manages an autofill context that the autofill service can later save user input from.  When `finishAutofillContext` is called, the text input plugin will first try to trigger the platform's autofill user input saving if requested, and destroys the current autofill context. 

### Why only one autofill context
Currently only one autofill context can exist at any given time. 

On Android it seems at most one `AutofillManager` instance can be created for each android `Context`. Each `AutofillManager` manages its own autofill context ("Session") and can't have more than one at the same time (however, different activities have separate `AutofillManager` instances so they have separate contexts). Since Flutter only has one `Acitivity`, we can't create a new autofill context on Android without finishing the current one.

This becomes a problem when the navigation scheme gets a bit complex. For example: 
- In a multi-tab app where each tab contains a `Form`, when we save we would have to save contents in all the tabs. I think a native android app would be in the same situation if it implements tab bar navigation using Fragments.
- In an autofillable `From` that changes the context temporarily (e.g., requires login to continue, and goes back to the `Form` when logged in). Saving the login credentials when the user finishes logging in would also save the unfinished form. Saving everything when the user submits the form is a bit misleading.

 For iOS and web we manage our own autofill contexts in the text input plugin. But having a separate context for each route doesn't sound easy because we would have nested contexts when there are nested navigators.
 

## Engine

https://github.com/flutter/engine/pull/18643

## Related Issues

https://github.com/flutter/flutter/issues/55613

## Tests

I added the following tests:

Tests will be added later.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

LongCatIsLooong committed:
WIP

LongCatIsLooong committed:
docs

LongCatIsLooong committed:
docs

skia-gold commented:
Gold has detected about 2 untriaged digest(s) on patchset 3.
Please triage them at https://flutter-gold.skia.org/cl/github/58731

justinmc commented on packages/flutter/lib/src/services/autofill.dart:
```diff
@@ -799,13 +799,36 @@ mixin AutofillScopeMixin implements AutofillScope {
+            || (hints?.contains(AutofillHints.newPassword) ?? false);
+        }
+
+        // In a native iOS app, Only password fields that have multi-field
```
Lowercase "o" on "Only" here.

Also I think you forgot a word after that. "save"?

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1179,6 +1180,7 @@ class TextInput {
     _channel.invokeMethod<void>('TextInput.requestAutofill');
   }
 
+
```
Extra empty line.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1192,4 +1194,76 @@ class TextInput {
     );
   }
+
+  /// Finishes the current autofill context, potentially saves the user input
```
"Finishes the current autofill context and potentially saves the user input"

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1192,4 +1194,76 @@ class TextInput {
+  /// {@endtemplate}
+  ///
+  /// Typically, this method should be called when the user has finalized their
+  /// input and ready to submit. For example, in a [Form], it's typically done
```
"input and is ready to submit."

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1192,4 +1194,76 @@ class TextInput {
+  /// [AutofillContextLifecycleAction] widget provides configurable event hooks
+  /// on [Navigator]s and [Form]s that you can call [finishAutofillContext] from.
+  /// By default, it calls [finishAutofillContext] with [shouldSave] equals to
+  /// `true` when a [Form] is saved, or whenever the an opaque [ModalRoute] in
```
Duplicate "the an" here.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1192,4 +1194,76 @@ class TextInput {
+  /// Currently there can be at most one autofill context, and
+  /// [finishAutofillContext] either saves/discards everything or nothing at all.
+  /// With more complex navigation schemes, the ability to switch autofill
+  /// contexts is desirable.
```
This feels a bit inconclusive.  Is there anything we can recommend for users to do who want multiple contexts?

justinmc commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -225,3 +228,142 @@ class _AutofillScope extends InheritedWidget {
   bool updateShouldNotify(_AutofillScope old) => _scope != old._scope;
 }
+
+/// An [InheritedWidget] that configures its descendant [Navigator]s and [Form]s,
```
Nit: I think there should be no comma at the end of this line, but it's fine either way.

justinmc commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -225,3 +228,142 @@ class _AutofillScope extends InheritedWidget {
+/// See also:
+///
+/// * [AutofillContextLifecycleDelegate], the configuration data this
+///   [InheritWidget] carries.
```
InheritWidget => InheritedWidget

justinmc commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -225,3 +228,142 @@ class _AutofillScope extends InheritedWidget {
+    this.lifecycleDelegate,
+  }): super(key: key, child: child);
+
+  /// The [AutofillContextLifecycleDelegate] that descendant
```
I think you meant to finish this sentence here.

justinmc commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -225,3 +228,142 @@ class _AutofillScope extends InheritedWidget {
+  /// The [AutofillContextLifecycleDelegate] that descendant
+  final AutofillContextLifecycleDelegate lifecycleDelegate;
+
+  /// The [AutofillContextLifecycleDelegate] to fall back to, when there's no
```
Nit: Another comma that I feel like should be removed.

justinmc commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -225,3 +228,142 @@ class _AutofillScope extends InheritedWidget {
+}
+
+/// Configures whether the current autofill context should be saved or discarded
+/// , or no action should be taken at all, when [Navigator]s and [Form]s
```
Weird line break here with a floating comma.

justinmc commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -225,3 +228,142 @@ class _AutofillScope extends InheritedWidget {
+  void didPush(Route<dynamic> route, Route<dynamic> previousRoute) {
+    // TODO(LongCatIsLooong): Ideally we want to set the current context aside
+    // and start a new autofill context for the new route we pushed. This
+    // doesn't seem to be possible on Android currently.
```
Maybe open an issue about this and link to it here.

justinmc reviewed (commented):
Mostly just a bunch of quick nits here.  I'm interested in what others think about the inherited widget calls in Navigator, but I can't think of a better approach.

chunhtai commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -2371,10 +2378,12 @@ class _RouteEntry extends RouteTransitionRecord {
     if (previousState == _RouteLifecycle.replace || previousState == _RouteLifecycle.pushReplace) {
       for (final NavigatorObserver observer in navigator.widget.observers)
         observer.didReplace(newRoute: route, oldRoute: previousPresent);
+      AutofillContextLifecycleAction.of(navigator.context).didReplace(newRoute: route, oldRoute: previousPresent);
```
Can we avoid this? we want to move away from using navigator observer-like system?

LongCatIsLooong replied:
Yes I would like to avoid doing this. Is there a better way to monitor global route changes? (Or did you mean `NavigatorObserver`s should not be used in general)?

chunhtai replied:
At some point we want to remove NavigatorObservers. It was a compromise to implement hero controller, and people started to use it to monitor the route history which is not what it built for. This result in people wants to add more and more functionality to this class and things started to get out of control. One other way will be creating a new route class to monitor Route.didpush, Route.didpop and such...

goderbauer replied:
It's very odd to special case this. Also, why does autofill depend on the navigator?

nturgut commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -225,3 +228,142 @@ class _AutofillScope extends InheritedWidget {
+  /// [savedForm].
+  ///
+  /// The default implementation saves the current autofill context.
+  void onFormSave(FormState savedForm) {
```
question: why does the function receive savedForm? I remember we discussed in the future we might want to send more parameters with save. Is it related to discussion?

LongCatIsLooong replied:
The `FormState` here is just a bit of additional information that helps the developer determine what needs to be done when this particular form gets saved. The framework doesn't really send the information to the engine.

nturgut commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -225,3 +228,142 @@ class _AutofillScope extends InheritedWidget {
+  ///
+  /// The default implementation does not do anything.
+  void didRemove(Route<dynamic> route, Route<dynamic> previousRoute) {
+  }
```
Should this call shouldSave: false?

nturgut commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -225,3 +228,142 @@ class _AutofillScope extends InheritedWidget {
+/// context (before discarding it), when any of the affected [Form]s is saved,
+/// or any of the affected [Navigator]s pushes/pops opaque [ModalRoute], or the
+/// current route is replaced with an opaque [ModalRoute].
+class AutofillContextLifecycleDelegate {
```
There is nothing in the lifecycle which calls "shouldSave: false", is it possible to have a scenario that we want to cancel the form?

nturgut commented on packages/flutter/lib/src/services/autofill.dart:
```diff
@@ -799,13 +799,36 @@ mixin AutofillScopeMixin implements AutofillScope {
+          final Iterable<String> hints = client.textInputConfiguration.autofillConfiguration?.autofillHints;
+          return client.textInputConfiguration.obscureText
+            || (hints?.contains(AutofillHints.password) ?? false)
+            || (hints?.contains(AutofillHints.newPassword) ?? false);
```
Question: why are we adding this check in the framework, can we handle it on the iOS engine side?

The reason that I'm concerned is, platform is does not necessarily mean that this is a flutter app running as an iOS mobile app. It can also be a Flutter Web app running on iOS Safari. Since web also return TargetPlatform.iOS, in this case, this branch won't work correctly on browsers on iOS, running Flutter for Web. The code is here: https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/foundation/_platform_web.dart#L20

LongCatIsLooong replied:
I'll check if `kIsWeb`.

goderbauer commented on packages/flutter/lib/src/services/autofill.dart:
```diff
@@ -799,13 +799,36 @@ mixin AutofillScopeMixin implements AutofillScope {
+            || (hints?.contains(AutofillHints.newPassword) ?? false);
+        }
+
+        // In a native iOS app, Only password fields that have multi-field
```
Can we just do this on the iOS specific implementation in the engine?

LongCatIsLooong replied:
It is less data we're sending to the engine on iOS I guess?

goderbauer commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -225,3 +228,142 @@ class _AutofillScope extends InheritedWidget {
+/// Set [AutofillContextLifecycleAction.globalLifecycleDelegate] if you wish to
+/// change the default [AutofillContextLifecycleDelegate] for the entire app.
+///
+/// The default [globalLifecycleDelegate] saves the current autofill context
```
Having this global is odd. Why do we need it? If you want different settings, you'd just wrap your subtree in a different `AutofillContextLifecycleAction` widget.

goderbauer commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -225,3 +228,142 @@ class _AutofillScope extends InheritedWidget {
+  /// The default implementation saves the current autofill context if [route]
+  /// is an opaque [ModalRoute].
+  void didPop(Route<dynamic> route, Route<dynamic> previousRoute) {
+    if (route is ModalRoute && route.opaque)
```
Why is this the condition for saving? What if I have my form an an Alert box or something like that?

goderbauer commented:
It is not super clear to me what we are trying to solve with this PR. In particular, why does "autofill save" get triggered by route changes? I would have expected that we only want to save if the user explicitly "commits" a textfield entry (e.g. by pressing the Login button). This may (but doesn't necessarily have to) cause a route change, but the signal to save isn't the route change. It's the user pressing a button to say "this data is valid now, send it".

LongCatIsLooong committed:
update

LongCatIsLooong committed:
docs and tests

LongCatIsLooong commented:
Updated. Does it look better now?

An edge case that I need to update the engine implementation for:
If a `TextField` doesn't have an `AutofillGroup` ancestor then it will not be added to the current autofill context (will be removed if it's already in the current autofill context). It does not change the current autofill behavior, as it will still be autofill (alone). On android its text value needs to be set to an empty string so it doesn't get saved.

I'll expand the `AutofillGroup` example to showcase the usage of the topmost group.

goderbauer commented on packages/flutter/lib/src/services/autofill.dart:
```diff
@@ -799,13 +799,36 @@ mixin AutofillScopeMixin implements AutofillScope {
-        currentClientConfiguration: configuration,
-      ),
+
+    TextInputConfiguration inputConfiguration;
```
Why do we need to do this filtering on the framework side? Can the iOS embedder just pick the fields that it finds relevant?

LongCatIsLooong replied:
I think by doing this we would be able to reduce the size of the messages we have to send to the platform every time the focus changes.

goderbauer commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -829,12 +829,13 @@ class TextInputConnection {
-  /// autofill, and the platform has a standalone autofill UI (for example, this
-  /// call has no effect on iOS since its autofill UI is part of the software
-  /// keyboard).
+  /// Currently only works on Android.
```
This should document what happens if called when not on Android.

goderbauer commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1192,4 +1193,62 @@ class TextInput {
   }
+
+  /// Finishes the current autofill context, and potentially saves the user
+  /// input for future use if [shouldSave] is true.
```
```suggestion
  /// input for future use if `shouldSave` is true.
```

goderbauer replied:
And elsewhere in this doc since it is a parameter to the method and not a member of the surrounding context that we can link to.

LongCatIsLooong replied:
It does get rendered differently I think, for instance `children` and `onValueChanged` on this page: https://api.flutter.dev/flutter/cupertino/CupertinoSlidingSegmentedControl/CupertinoSlidingSegmentedControl.html. Also the LSP client I'm using is able to pick up method parameters that are in square brackets.

goderbauer replied:
See also: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#dartdoc-specific-requirements

goderbauer commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -109,13 +140,19 @@ export 'package:flutter/services.dart' show AutofillHints;
   const AutofillGroup({
     Key key,
     @required this.child,
+    this.onDisposeAction = AutofillContextAction.commit,
```
assert that this is not null?

LongCatIsLooong replied:
It's nullable, and on dispose action is going to be a no-op when it's null.

goderbauer replied:
We should document what no-op in that case means (and instead of using null, can we define a `AutofillContextAction ` for that?

LongCatIsLooong replied:
I wonder why is a separate enum case preferred over `null`? I think in this case using null to stand for "no action should be taken" makes sense too. Is it because the analyzer has exhaustiveness check for enum cases?

goderbauer reviewed (commented):
Approach looks good.

Maybe @justinmc can take a closer look at the textfield integration.

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -1216,6 +1219,7 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       _currentAutofillScope?.unregister(autofillId);
       _currentAutofillScope = newAutofillGroup;
       newAutofillGroup?.register(this);
+      _isInAutofillContext = _isInAutofillContext || _shouldBeInAutofillContext;
```
_isInAutofillContext can never be set to false once it becomes true, is that right?

LongCatIsLooong replied:
Yes. If a text field's was in the current autofill context, and then its autofill gets disabled (by setting `autofillHints` to null), the engine needs to remove it from the autofill context. It lets the engine know by sending its new `TextInputConfiguration`, instead of remain silent in which case the engine will just assume the text field is gone and it will keep the old user input.

justinmc reviewed (approved):
Just a question below, otherwise the EditableText stuff looks good to me.

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/master' into autofill-save

LongCatIsLooong committed:
review

HansMuller commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -137,6 +174,18 @@ class AutofillGroup extends StatefulWidget {
+  ///
+  /// Defaults to [AutofillContextAction.commit], which prompts the platform to
+  /// save the user input and destroy the current autofill context. No
+  /// autofill context related actions will be automatically taken when
```
autofill context related actions will be automatically taken  => no action will be taken

HansMuller reviewed (commented):
Please update the PRs description: remove the last section about changing the navigator.

nturgut commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1213,4 +1215,62 @@ class TextInput {
+  /// (which is done automatically when an autofillable [EditableText] gains
+  /// focus), the current autofill context will merge the content of that
+  /// [AutofillGroup] into itself. When there isn't an existing autofill context,
+  /// one will be created to hold the newly added input fields from the group.
```
Question: Since we have only one `autofill context` multiple AutofillGroups can end up in the same context right?

LongCatIsLooong replied:
Yes

nturgut commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1213,4 +1215,62 @@ class TextInput {
+  /// the input fields while it's being processed.
+  ///
+  /// The topmost [AutofillGroup]s also have the ability to call
+  /// [finishAutofillContext] automatically when they are getting disposed. The
```
question: If "onDisposeAction" is called on an autofill group, and if it calls "finishAutofillContext", it will still save all the autofill values, is it correct?

LongCatIsLooong replied:
I guess it depends on what the `onDisposeAction` was. The `onDisposeAction`, if not null, is usually destructive. So if a finishAutofillContext call to save the user input is called after an `onDisposeAction`, that call has nothing to save because the context is destroyed at that point.

nturgut commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1213,4 +1215,62 @@ class TextInput {
+  /// exact behavior can be configured using [AutofillGroup.onDisposeAction].
+  ///
+  /// Calling [finishAutofillContext] may cause the platform to show the "save
+  /// for autofill" dialog and disrupts the user's flow. Ideally the dialog
```
Maybe we can reword this part as for example on platform X, "savefor autofill" dialog and disrupts the user's flow. Otherwise it might look like all platforms have a dialog with the same warning message, which might confuse the developer.

nturgut commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1213,4 +1215,62 @@ class TextInput {
+      shouldSave
+        ? 'TextInput.AutofillContext.commit'
+        : 'TextInput.AutofillContext.cancel',
+    );
```
Can we also change this method to send "commit/cancel" information as an argument to make it similar to the other methods?

LongCatIsLooong replied:
Update to take a bool. Let me update the engine PR.

nturgut reviewed (commented):
Thanks for working on this PR :)

I need to add a methods to the engine before you merge to PR, so that there are no exceptions on the web platform. I'll do it once we agree on the arguments/naming.

LongCatIsLooong committed:
Merge branch 'master' of github.com:flutter/flutter into autofill-save

LongCatIsLooong committed:
Merge branch 'master' of github.com:flutter/flutter into autofill-save

LongCatIsLooong committed:
update

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1213,4 +1215,62 @@ class TextInput {
+  /// Typically, this method should be called when the user has finalized their
+  /// input. For example, in a [Form], it's typically done immediately before or
+  /// after its content is submitted. You should make sure that any connected
+  /// [TextInputClient] is disconnected and no input field is focused before
```
removed the assert so it's no longer a hard requirement.

LongCatIsLooong committed:
fix tests

LongCatIsLooong commented:
@justinmc @goderbauer @nturgut @HansMuller addressed most comments (except the ones I still have questions for). Also the engine PR for iOS and Android is ready for review too: https://github.com/flutter/engine/pull/18643 (I made some non-trivial changes after the PR was approved. The CI failure seems irrelevant).

justinmc reviewed (approved):
LGTM with the changes 👍

nturgut commented on packages/flutter/lib/src/widgets/autofill.dart:
```diff
@@ -160,6 +208,8 @@ class AutofillGroup extends StatefulWidget {
 class AutofillGroupState extends State<AutofillGroup> with AutofillScopeMixin {
   final Map<String, AutofillClient> _clients = <String, AutofillClient>{};
 
+  bool _isTopmostAutofillGroup = false;
```
The definition on topmost (the ones that are closest to the root widget) is in the comments on line 54. It is not very close to this variable. In order to increase readability of the code, and make it easier to understand by developers checking this variable, let's document `topmost` here as well.

LongCatIsLooong replied:
Added a comment in the code.

LongCatIsLooong committed:
add comment to isTopmostAutofillGroup

LongCatIsLooong committed:
use existing matcher
